In [1]:
import os

# Set the path
folder_path = '/Users/emilsafarov/Library/CloudStorage/OneDrive-Personal/CF/A6/1_sourceing_open_data/US project/Data/Input/file_7_rent'

# Get all files in the folder
files = [f for f in os.listdir(folder_path) if os.path.isfile(os.path.join(folder_path, f))]

# Print file names
for file in files:
    print(file)


fips.xlsx
.DS_Store
fmr_2021.xlsx
fmr_2020.xlsx
fmr_2023.xlsx
fmr_2022.xlsx
merged_rent_data.xlsx
fmr_2024.xlsx
fmr_merged_all_years.csv


In [2]:
import os
import re
import pandas as pd

# 📁 Set your folder path here
folder_path = '/Users/emilsafarov/Library/CloudStorage/OneDrive-Personal/CF/A6/1_sourceing_open_data/US project/Data/Input/file_7_rent'

# 📄 List only .xlsx files
excel_files = [f for f in os.listdir(folder_path) if f.endswith('.xlsx')]

merged_df = None

for file in excel_files:
    file_path = os.path.join(folder_path, file)
    df = pd.read_excel(file_path)

    # 🗓️ Extract year from filename (e.g., "fmr_2020.xlsx")
    year_match = re.search(r'(\d{4})', file)
    if not year_match:
        print(f"⚠️ Skipping file (no year found): {file}")
        continue
    year = year_match.group(1)

    # 🏷️ Rename columns to include year
    df = df.rename(columns={
        col: f"{col}_{year}" for col in df.columns if col not in ['FIPS', 'State', 'County']
    })

    # 🔗 Merge on key columns
    if merged_df is None:
        merged_df = df
    else:
        merged_df = pd.merge(merged_df, df, on=['FIPS', 'State', 'County'], how='outer')

# ✅ Done — Preview and export
print("✅ Merged shape:", merged_df.shape)
print(merged_df.head())

# 💾 Save to CSV
output_path = os.path.join(folder_path, "fmr_merged_all_years.csv")
merged_df.to_csv(output_path, index=False)
print(f"✅ File saved to: {output_path}")

⚠️ Skipping file (no year found): fips.xlsx
⚠️ Skipping file (no year found): merged_rent_data.xlsx
✅ Merged shape: (4775, 28)
        FIPS  fmr_0_2021  fmr_1_2021  fmr_2_2021  fmr_3_2021  fmr_4_2021  \
0  100199999       640.0       766.0       908.0      1148.0      1520.0   
1  100399999       718.0       723.0       922.0      1249.0      1584.0   
2  100599999       488.0       492.0       648.0       806.0       907.0   
3  100799999       817.0       871.0      1002.0      1303.0      1409.0   
4  100999999       817.0       871.0      1002.0      1303.0      1409.0   

           County State  fmr_0_2020  fmr_1_2020  ...  fmr_0_2022  fmr_1_2022  \
0  Autauga County    AL       583.0       702.0  ...       643.0       764.0   
1  Baldwin County    AL       744.0       749.0  ...       772.0       777.0   
2  Barbour County    AL       477.0       481.0  ...       532.0       536.0   
3     Bibb County    AL       804.0       861.0  ...       765.0       820.0   
4   Blount Count

In [3]:
merged_df

FIPS  fmr_0_2021  fmr_1_2021  fmr_2_2021  fmr_3_2021  fmr_4_2021  \
0      100199999       640.0       766.0       908.0      1148.0      1520.0   
1      100399999       718.0       723.0       922.0      1249.0      1584.0   
2      100599999       488.0       492.0       648.0       806.0       907.0   
3      100799999       817.0       871.0      1002.0      1303.0      1409.0   
4      100999999       817.0       871.0      1002.0      1303.0      1409.0   
...          ...         ...         ...         ...         ...         ...   
4770  7215199999       436.0       477.0       556.0       740.0       909.0   
4771  7215399999       361.0       369.0       421.0       596.0       630.0   
4772  7801099999       835.0       852.0      1040.0      1294.0      1410.0   
4773  7802099999      1109.0      1314.0      1633.0      2031.0      2214.0   
4774  7803099999       797.0       962.0      1204.0      1498.0      1633.0   

                 County State  fmr_0_2020  fmr_1_2020  ...  fmr_0_2022  \
0        Autauga County    AL       583.0       702.0  ...       643.0   
1        Baldwin County    AL       744.0       749.0  ...       772.0   
2        Barbour County    AL       477.0       481.0  ...       532.0   
3           Bibb County    AL       804.0       861.0  ...       765.0   
4         Blount County    AL       804.0       861.0  ...       765.0   
...                 ...   ...         ...         ...  ...         ...   
4770  Yabucoa Municipio    PR       416.0       451.0  ...       428.0   
4771    Yauco Municipio    PR       349.0       359.0  ...       362.0   
4772          St. Croix    VI       799.0       815.0  ...       868.0   
4773           St. John    VI      1061.0      1258.0  ...      1154.0   
4774         St. Thomas    VI       763.0       921.0  ...       832.0   

      fmr_1_2022  fmr_2_2022  fmr_3_2022  fmr_4_2022  fmr_0_2024  fmr_1_2024  \
0          764.0       914.0      1156.0      1494.0       836.0       913.0   
1          777.0      1017.0      1348.0      1715.0      1051.0      1056.0   
2          536.0       705.0       871.0       980.0       652.0       656.0   
3          820.0       943.0      1220.0      1316.0       983.0      1109.0   
4          820.0       943.0      1220.0      1316.0       983.0      1109.0   
...          ...         ...         ...         ...         ...         ...   
4770       466.0       538.0       719.0       868.0       556.0       583.0   
4771       370.0       422.0       572.0       574.0       410.0       421.0   
4772       886.0      1082.0      1338.0      1467.0      1084.0      1106.0   
4773      1368.0      1700.0      2101.0      2305.0      1441.0      1707.0   
4774      1001.0      1253.0      1549.0      1699.0      1059.0      1249.0   

      fmr_2_2024  fmr_3_2024  fmr_4_2024  
0         1092.0      1383.0      1753.0  
1         1362.0      1670.0      2114.0  
2          857.0      1089.0      1141.0  
3         1245.0      1570.0      1752.0  
4         1245.0      1570.0      1752.0  
...          ...         ...         ...  
4770       668.0       882.0      1056.0  
4771       473.0       628.0       630.0  
4772      1351.0      1634.0      1799.0  
4773      2122.0      2566.0      2826.0  
4774      1563.0      1890.0      2082.0  

[4775 rows x 28 columns]

In [4]:
# ✂️ Trim last 5 digits from FIPS to get standard 5-digit county codes
merged_df['FIPS'] = df['FIPS'].astype(str).str[:-5]


In [5]:
merged_df

FIPS  fmr_0_2021  fmr_1_2021  fmr_2_2021  fmr_3_2021  fmr_4_2021  \
0     1001       640.0       766.0       908.0      1148.0      1520.0   
1     1003       718.0       723.0       922.0      1249.0      1584.0   
2     1005       488.0       492.0       648.0       806.0       907.0   
3     1007       817.0       871.0      1002.0      1303.0      1409.0   
4     1009       817.0       871.0      1002.0      1303.0      1409.0   
...    ...         ...         ...         ...         ...         ...   
4770   NaN       436.0       477.0       556.0       740.0       909.0   
4771   NaN       361.0       369.0       421.0       596.0       630.0   
4772   NaN       835.0       852.0      1040.0      1294.0      1410.0   
4773   NaN      1109.0      1314.0      1633.0      2031.0      2214.0   
4774   NaN       797.0       962.0      1204.0      1498.0      1633.0   

                 County State  fmr_0_2020  fmr_1_2020  ...  fmr_0_2022  \
0        Autauga County    AL       583.0       702.0  ...       643.0   
1        Baldwin County    AL       744.0       749.0  ...       772.0   
2        Barbour County    AL       477.0       481.0  ...       532.0   
3           Bibb County    AL       804.0       861.0  ...       765.0   
4         Blount County    AL       804.0       861.0  ...       765.0   
...                 ...   ...         ...         ...  ...         ...   
4770  Yabucoa Municipio    PR       416.0       451.0  ...       428.0   
4771    Yauco Municipio    PR       349.0       359.0  ...       362.0   
4772          St. Croix    VI       799.0       815.0  ...       868.0   
4773           St. John    VI      1061.0      1258.0  ...      1154.0   
4774         St. Thomas    VI       763.0       921.0  ...       832.0   

      fmr_1_2022  fmr_2_2022  fmr_3_2022  fmr_4_2022  fmr_0_2024  fmr_1_2024  \
0          764.0       914.0      1156.0      1494.0       836.0       913.0   
1          777.0      1017.0      1348.0      1715.0      1051.0      1056.0   
2          536.0       705.0       871.0       980.0       652.0       656.0   
3          820.0       943.0      1220.0      1316.0       983.0      1109.0   
4          820.0       943.0      1220.0      1316.0       983.0      1109.0   
...          ...         ...         ...         ...         ...         ...   
4770       466.0       538.0       719.0       868.0       556.0       583.0   
4771       370.0       422.0       572.0       574.0       410.0       421.0   
4772       886.0      1082.0      1338.0      1467.0      1084.0      1106.0   
4773      1368.0      1700.0      2101.0      2305.0      1441.0      1707.0   
4774      1001.0      1253.0      1549.0      1699.0      1059.0      1249.0   

      fmr_2_2024  fmr_3_2024  fmr_4_2024  
0         1092.0      1383.0      1753.0  
1         1362.0      1670.0      2114.0  
2          857.0      1089.0      1141.0  
3         1245.0      1570.0      1752.0  
4         1245.0      1570.0      1752.0  
...          ...         ...         ...  
4770       668.0       882.0      1056.0  
4771       473.0       628.0       630.0  
4772      1351.0      1634.0      1799.0  
4773      2122.0      2566.0      2826.0  
4774      1563.0      1890.0      2082.0  

[4775 rows x 28 columns]

In [6]:
# 🔍 Find duplicated rows in merged_df where FIPS, State, and County are the same
duplicates = merged_df[merged_df.duplicated(subset=['FIPS', 'State', 'County'], keep=False)]

# Show number and preview
print(f"⚠️ Number of duplicated rows: {duplicates.shape[0]}")
print(duplicates.sort_values(by=['FIPS', 'State', 'County']))


⚠️ Number of duplicated rows: 1551
       FIPS  fmr_0_2021  fmr_1_2021  fmr_2_2021  fmr_3_2021  fmr_4_2021  \
1340  23001       685.0       737.0       947.0      1214.0      1484.0   
1341  23001       685.0       737.0       947.0      1214.0      1484.0   
1342  23001       685.0       737.0       947.0      1214.0      1484.0   
1343  23001       685.0       737.0       947.0      1214.0      1484.0   
1344  23001       685.0       737.0       947.0      1214.0      1484.0   
...     ...         ...         ...         ...         ...         ...   
473    9015       816.0       818.0      1036.0      1307.0      1695.0   
474    9015       816.0       818.0      1036.0      1307.0      1695.0   
475    9015       816.0       818.0      1036.0      1307.0      1695.0   
476    9015       816.0       818.0      1036.0      1307.0      1695.0   
477    9015       816.0       818.0      1036.0      1307.0      1695.0   

                   County State  fmr_0_2020  fmr_1_2020  ...  fm

In [7]:
# 🔍 Count and drop fully duplicated rows (every column the same)
num_duplicates = merged_df.duplicated().sum()
print(f"⚠️ Number of fully duplicated rows dropped: {num_duplicates}")

# 🧹 Drop full duplicates
merged_df = merged_df.drop_duplicates()


⚠️ Number of fully duplicated rows dropped: 1388


In [8]:
# Re-check for partial duplicates (same FIPS, State, County)
duplicates = merged_df[merged_df.duplicated(subset=['FIPS', 'State', 'County'], keep=False)]

print(f"🔁 Remaining duplicated FIPS/State/County rows: {duplicates.shape[0]}")


🔁 Remaining duplicated FIPS/State/County rows: 81


In [9]:
# 🧠 Group by FIPS, State, County and take the median for numeric columns
merged_df = merged_df.groupby(['FIPS', 'State', 'County'], as_index=False).median(numeric_only=True)


In [10]:
merged_df

FIPS State             County  fmr_0_2021  fmr_1_2021  fmr_2_2021  \
0     10001    CT     Windham County       816.0       818.0      1036.0   
1     10003    DE        Kent County       878.0       908.0      1066.0   
2     10005    DE  New Castle County       900.0      1040.0      1260.0   
3      1001    AL     Autauga County       640.0       766.0       908.0   
4      1003    AL     Baldwin County       718.0       723.0       922.0   
...     ...   ...                ...         ...         ...         ...   
3319   9011    CT  New London County       828.0       991.5      1254.5   
3320   9013    CT  New London County       810.0       976.0      1227.0   
3321   9013    CT     Tolland County       889.0      1091.0      1347.0   
3322   9015    CT     Tolland County       889.0      1091.0      1347.0   
3323   9015    CT     Windham County       816.0       818.0      1036.0   

      fmr_3_2021  fmr_4_2021  fmr_0_2020  fmr_1_2020  ...  fmr_0_2022  \
0         1307.0      1695.0       787.0       811.0  ...       848.0   
1         1526.0      1785.0       880.0       893.0  ...       851.0   
2         1567.0      1796.0       864.0      1013.0  ...       940.0   
3         1148.0      1520.0       583.0       702.0  ...       643.0   
4         1249.0      1584.0       744.0       749.0  ...       772.0   
...          ...         ...         ...         ...  ...         ...   
3319      1709.5      2123.5       826.5       953.5  ...       854.5   
3320      1584.0      2027.0       815.0       938.0  ...       833.0   
3321      1675.0      1958.0       801.0       993.0  ...       865.0   
3322      1675.0      1958.0       801.0       993.0  ...       865.0   
3323      1307.0      1695.0       787.0       811.0  ...       848.0   

      fmr_1_2022  fmr_2_2022  fmr_3_2022  fmr_4_2022  fmr_0_2024  fmr_1_2024  \
0          854.0      1110.0      1388.0      1875.0      1186.0      1194.0   
1          856.0      1002.0      1423.0      1647.0      1110.0      1117.0   
2         1071.0      1298.0      1605.0      1837.0      1303.0      1451.0   
3          764.0       914.0      1156.0      1494.0       836.0       913.0   
4          777.0      1017.0      1348.0      1715.0      1051.0      1056.0   
...          ...         ...         ...         ...         ...         ...   
3319      1020.5      1286.0      1742.5      2179.0      1102.5      1277.0   
3320      1006.0      1254.0      1616.0      2102.0      1055.0      1263.0   
3321      1054.0      1302.0      1609.0      1907.0      1121.0      1319.0   
3322      1054.0      1302.0      1609.0      1907.0      1121.0      1319.0   
3323       854.0      1110.0      1388.0      1875.0      1186.0      1194.0   

      fmr_2_2024  fmr_3_2024  fmr_4_2024  
0         1569.0      1921.0      2380.0  
1         1368.0      1928.0      2152.0  
2         1737.0      2100.0      2422.0  
3         1092.0      1383.0      1753.0  
4         1362.0      1670.0      2114.0  
...          ...         ...         ...  
3319      1626.0      2134.5      2740.5  
3320      1556.0      2030.0      2625.0  
3321      1654.0      2000.0      2425.0  
3322      1654.0      2000.0      2425.0  
3323      1569.0      1921.0      2380.0  

[3324 rows x 28 columns]

In [11]:
# 🗂️ Print all unique state names
unique_states = sorted(merged_df['State'].dropna().unique())

print("📍 Unique States:")
for state in unique_states:
    print(state)


📍 Unique States:
AK
AL
AR
AS
AZ
CA
CO
CT
DC
DE
FL
GA
GU
HI
IA
ID
IL
IN
KS
KY
LA
MA
MD
ME
MI
MN
MO
MP
MS
MT
NC
ND
NE
NH
NJ
NM
NV
NY
OH
OK
OR
PA
PR
RI
SC
SD
TN
TX
UT
VA
VT
WA
WI
WV
WY


In [12]:
# 🧹 Remove U.S. territories — keep only the 50 states + DC
exclude_territories = ['AS', 'GU', 'MP', 'PR', 'VI']
merged_df = merged_df[~merged_df['State'].isin(exclude_territories)]


In [13]:
# 🔧 Pad FIPS codes with leading zeros where length < 5
merged_df['FIPS'] = merged_df['FIPS'].astype(str).str.zfill(5)

# 👀 Preview the updated DataFrame
merged_df

/var/folders/tt/mzspvn2d7hb3qsps4c4lqh0c0000gn/T/ipykernel_15349/1647850853.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df['FIPS'] = merged_df['FIPS'].astype(str).str.zfill(5)


FIPS State             County  fmr_0_2021  fmr_1_2021  fmr_2_2021  \
0     10001    CT     Windham County       816.0       818.0      1036.0   
1     10003    DE        Kent County       878.0       908.0      1066.0   
2     10005    DE  New Castle County       900.0      1040.0      1260.0   
3     01001    AL     Autauga County       640.0       766.0       908.0   
4     01003    AL     Baldwin County       718.0       723.0       922.0   
...     ...   ...                ...         ...         ...         ...   
3319  09011    CT  New London County       828.0       991.5      1254.5   
3320  09013    CT  New London County       810.0       976.0      1227.0   
3321  09013    CT     Tolland County       889.0      1091.0      1347.0   
3322  09015    CT     Tolland County       889.0      1091.0      1347.0   
3323  09015    CT     Windham County       816.0       818.0      1036.0   

      fmr_3_2021  fmr_4_2021  fmr_0_2020  fmr_1_2020  ...  fmr_0_2022  \
0         1307.0      1695.0       787.0       811.0  ...       848.0   
1         1526.0      1785.0       880.0       893.0  ...       851.0   
2         1567.0      1796.0       864.0      1013.0  ...       940.0   
3         1148.0      1520.0       583.0       702.0  ...       643.0   
4         1249.0      1584.0       744.0       749.0  ...       772.0   
...          ...         ...         ...         ...  ...         ...   
3319      1709.5      2123.5       826.5       953.5  ...       854.5   
3320      1584.0      2027.0       815.0       938.0  ...       833.0   
3321      1675.0      1958.0       801.0       993.0  ...       865.0   
3322      1675.0      1958.0       801.0       993.0  ...       865.0   
3323      1307.0      1695.0       787.0       811.0  ...       848.0   

      fmr_1_2022  fmr_2_2022  fmr_3_2022  fmr_4_2022  fmr_0_2024  fmr_1_2024  \
0          854.0      1110.0      1388.0      1875.0      1186.0      1194.0   
1          856.0      1002.0      1423.0      1647.0      1110.0      1117.0   
2         1071.0      1298.0      1605.0      1837.0      1303.0      1451.0   
3          764.0       914.0      1156.0      1494.0       836.0       913.0   
4          777.0      1017.0      1348.0      1715.0      1051.0      1056.0   
...          ...         ...         ...         ...         ...         ...   
3319      1020.5      1286.0      1742.5      2179.0      1102.5      1277.0   
3320      1006.0      1254.0      1616.0      2102.0      1055.0      1263.0   
3321      1054.0      1302.0      1609.0      1907.0      1121.0      1319.0   
3322      1054.0      1302.0      1609.0      1907.0      1121.0      1319.0   
3323       854.0      1110.0      1388.0      1875.0      1186.0      1194.0   

      fmr_2_2024  fmr_3_2024  fmr_4_2024  
0         1569.0      1921.0      2380.0  
1         1368.0      1928.0      2152.0  
2         1737.0      2100.0      2422.0  
3         1092.0      1383.0      1753.0  
4         1362.0      1670.0      2114.0  
...          ...         ...         ...  
3319      1626.0      2134.5      2740.5  
3320      1556.0      2030.0      2625.0  
3321      1654.0      2000.0      2425.0  
3322      1654.0      2000.0      2425.0  
3323      1569.0      1921.0      2380.0  

[3251 rows x 28 columns]

In [14]:
# 🔁 Check for duplicates by FIPS, State, County
dup_count = merged_df.duplicated(subset=['FIPS', 'State', 'County']).sum()
print(f"🔁 Duplicated rows based on FIPS, State, County: {dup_count}")

# 📊 Check for missing values
missing_summary = merged_df.isna().sum()
missing_summary = missing_summary[missing_summary > 0]

print("\n🔍 Missing values summary:")
print(missing_summary)

# 📦 Show total number of rows
print(f"\n📦 Total rows after cleaning: {merged_df.shape[0]}")


🔁 Duplicated rows based on FIPS, State, County: 0

🔍 Missing values summary:
fmr_0_2021    2
fmr_1_2021    2
fmr_2_2021    2
fmr_3_2021    2
fmr_4_2021    2
fmr_0_2020    2
fmr_1_2020    2
fmr_2_2020    2
fmr_3_2020    2
fmr_4_2020    2
fmr_0_2023    1
fmr_1_2023    1
fmr_2_2023    1
fmr_3_2023    1
fmr_4_2023    1
fmr_0_2022    2
fmr_1_2022    2
fmr_2_2022    2
fmr_3_2022    2
fmr_4_2022    2
fmr_0_2024    1
fmr_1_2024    1
fmr_2_2024    1
fmr_3_2024    1
fmr_4_2024    1
dtype: int64

📦 Total rows after cleaning: 3251


In [15]:
# 🧹 Filter rows with any missing values
missing_rows = merged_df[merged_df.isna().any(axis=1)]

# 📊 Show how many rows have missing values
print(f"🔍 Total rows with missing values: {missing_rows.shape[0]}")

# 👀 Preview the missing rows
print(missing_rows)


🔍 Total rows with missing values: 3
      FIPS State                      County  fmr_0_2021  fmr_1_2021  \
749  02063    AK         Chugach Census Area         NaN         NaN   
750  02066    AK    Copper River Census Area         NaN         NaN   
955  02275    AK  Valdez-Cordova Census Area       740.0       983.0   

     fmr_2_2021  fmr_3_2021  fmr_4_2021  fmr_0_2020  fmr_1_2020  ...  \
749         NaN         NaN         NaN         NaN         NaN  ...   
750         NaN         NaN         NaN         NaN         NaN  ...   
955      1121.0      1394.0      1675.0       809.0      1090.0  ...   

     fmr_0_2022  fmr_1_2022  fmr_2_2022  fmr_3_2022  fmr_4_2022  fmr_0_2024  \
749         NaN         NaN         NaN         NaN         NaN       989.0   
750         NaN         NaN         NaN         NaN         NaN       832.0   
955       764.0       987.0      1150.0      1422.0      1968.0         NaN   

     fmr_1_2024  fmr_2_2024  fmr_3_2024  fmr_4_2024  
749      1077.0

In [16]:
# 🎯 Goal: For each year-based numeric column, fill missing values using the mean for Alaska (State == 'AK')

import numpy as np

for col in merged_df.columns:
    if col not in ['FIPS', 'State', 'County'] and merged_df[col].dtype in [np.float64, np.int64]:
        # Calculate the mean for Alaska in the current column
        ak_mean = merged_df.loc[merged_df['State'] == 'AK', col].mean(skipna=True)

        # Fill missing values in this column with the Alaska mean
        merged_df[col] = merged_df[col].fillna(ak_mean)


/var/folders/tt/mzspvn2d7hb3qsps4c4lqh0c0000gn/T/ipykernel_15349/1227508098.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df[col] = merged_df[col].fillna(ak_mean)


🧼 Final Data Cleaning Report

Dataset: County-level data (including FIPS, State, County)
Goal: Prepare a clean, deduplicated, U.S.-only dataset ready for analysis

✅ Steps Performed
1. Removed U.S. Territories

Excluded rows where State was one of the following:
AS (American Samoa), GU (Guam), MP (Northern Mariana Islands), PR (Puerto Rico), VI (U.S. Virgin Islands)
✅ Retained DC (District of Columbia) as a valid county-equivalent
2. FIPS Code Cleanup

Removed trailing digits from inflated FIPS-like values (e.g., 2300177800 → 23001)
Ensured all FIPS values were properly padded to 5-digit strings (str.zfill(5))
Verified proper FIPS structure: 2-digit state code + 3-digit county code
3. Deduplication

✅ Step 1: Dropped fully duplicated rows (identical across all columns)
✅ Step 2: Removed partial duplicates based on FIPS, State, and County
Grouped by these fields
Aggregated numeric values using the median
Ensured one clean row per county
4. Missing Value Handling

Detected and handled missing values using state-based imputation
For each numeric column, missing values were filled with the mean of Alaska (AK) for the corresponding column
This approach preserved the structure and comparability of the data
✅ Final dataset contains no missing (NaN) values

In [20]:
# Set export path with corrected file name
export_path = "/Users/emilsafarov/Library/CloudStorage/OneDrive-Personal/CF/A6/1_sourceing_open_data/US project/Data/Output/Processed_datasets/confined_fmr_v2.xlsx"

# Export to Excel
merged_df.to_excel(export_path, index=False)

print(f"✅ Exported as: {export_path}")


✅ Exported as: /Users/emilsafarov/Library/CloudStorage/OneDrive-Personal/CF/A6/1_sourceing_open_data/US project/Data/Output/Processed_datasets/confined_fmr_v2.xlsx
